source/connect/connection-options/server-selection.txt

```diff
@@ -1,68 +1,195 @@
+
+.. literalinclude:: /includes/fundamentals/code-examples/connection/ServerSelection.cs
+   :start-after: start-server-selector
+   :end-before: end-server-selector
+   :language: csharp
+   :dedent:
+
+To learn more about the different server selector classes, see the
+`ServerSelectors API documentation <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.Core.Clusters.ServerSelectors.html>`__.
+
+Implementing Custom Server Selection Logic
+------------------------------------------
+
+You can implement your own custom server selection logic by creating a class that
+inherits from the ``IServerSelector`` interface and overrides the
+``SelectServers()`` method. The following example shows a simple custom server
+selection class that selects servers with a ``ServerType`` of
+``ServerType.ReplicaSetSecondary``:
+
+.. literalinclude:: /includes/fundamentals/code-examples/connection/ServerSelection.cs
+   :start-after: start-custom-class
+   :end-before: end-custom-class
+   :language: csharp
+   :dedent:
+
+You can then pass an instance of this class to the ``PreServerSelector`` or
+``PostServerSelector`` property of a ``ClusterConfigurator`` instance, as shown in the
+:ref:`csharp-server-selection-algorithm` section.
+
+Using Settings to Configure Server Selection
+--------------------------------------------
+
+You can specify the following server selection settings in your ``MongoClient`` object or
+in your connection URI:
+
+.. list-table::
+   :widths: 30 70
+   :header-rows: 1
+
+   * - Setting
+     - Description
+
+   * - ``LocalThreshold``
+     - | The latency window for server eligibility. If a server's round trip takes longer
+       | than the fastest server's round-trip time plus this value, the server isn't
+       | eligible for selection.
+       |
+       | **Data Type**: ``TimeSpan``
+       | **Default**: 15 milliseconds
+       | **Connection URI Example**: ``localThresholdMS=0``
+       
+   * - ``ReadPreference``
+     - | The client's default read-preference settings. ``MaxStaleness`` represents the
+       | longest replication lag (in real time) that a secondary can experience and
+       | still be eligible for server selection. Specifying ``-1`` means no maximum.
+       | See :ref:`read preference <read-preference>` for more information.
+       |
+       | **Data Type**: `ReadPreference <{+new-api-root+}/MongoDB.Driver/MongoDB.Driver.ReadPreference.html>`__
+       | **Default**: ``ReadPreference.Primary``
+       | **Connection URI Example**:
+
+       .. code-block:: none
+          :copyable: false
+
+          readPreference=primaryPreferred
+          &maxStalenessSeconds=90
+          &readPreferenceTags=dc:ny,rack:1
+
+   * - ``ServerSelectionTimeout``
+     - | The length of time the driver tries to select a server before timing out.
+       |
+       | **Data Type**: ``TimeSpan``
+       | **Default**: 30 seconds
+       | **Connection URI Example**: ``serverSelectionTimeoutMS=15000``
```

Comment:
Should we add wait queue settings here as well? So documentation cover the situation when there are no know eligible servers for the moment and users know how to tweak such cases as well.